test/controllers/tasks_controller_test.rb

```diff
@@ -105,24 +107,97 @@
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
     it "can update an existing task" do
-      skip
-      # Your code here
+      task_hash = {
+        task: {
+          name: "new task",
+          description: "new task description",
+          completed: false,
+        },
+      }
+
+      task = Task.create(name: "Stuff", description: "Lots of it!")
+
+      # Act-Assert
+      expect {
+        patch task_path(task.id), params: task_hash
+      }.wont_change "Task.count"
+
+      updated_task = Task.find_by(id: task.id)
+      expect(updated_task.name).must_equal task_hash[:task][:name]
+      expect(updated_task.description).must_equal task_hash[:task][:description]
+      expect(updated_task.completed).must_equal task_hash[:task][:completed]
+
+      must_respond_with :redirect
+      must_redirect_to task_path(task.id)
     end
 
-    it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+    it "will redirect if given an invalid id" do
+      task_hash = {
+        task: {
+          name: "new task",
+          description: "new task description",
+          completed: false,
+        },
+      }
+
+      invalid_id = -1
+
+      expect {
+        patch task_path(invalid_id), params: task_hash
+      }.wont_change "Task.count"
+
+      must_redirect_to tasks_path
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "can delete a task" do
+      task = Task.create(name: "thing", description: "yup")
+
+      expect {
+        delete task_path(task.id)
+      }.must_change "Task.count", -1
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+    end
+
+    it "will redirect if task can't be found" do
+      invalid_id = -1
+
+      expect {
+        delete task_path(invalid_id)
+      }.wont_change "Task.count"
 
+      must_redirect_to tasks_path
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "can mark a task complete" do
+      task = Task.create(name: "Thing", description: "Thing Thing")
+
+      expect(task.completed).must_equal false
+
+      expect {
+        patch mark_complete_path(task.id)
+      }.wont_change "Task.count"
+
+      expect(task.reload.completed).must_equal true
+      patch mark_complete_path(task.id)
+      expect(task.reload.completed).must_equal false
+    end
+
+    it "can update a task with a complete date/time" do
+      task = Task.create(name: "Thing", description: "Thing Thing")
+
+      expect(task.completion_date).must_equal nil
+
+      patch mark_complete_path(task.id)
+
+      expect(task.reload.completion_date).must_equal Date.today
```

Comment:
You should also write a test that you can toggle the task back to incomplete after being marked complete.